Lockbox Recursive Dependency Update

davidboren opened the pull request:
Okay!  This thing is ready for code review, though probably not merging.  Most of the relevant code is found in dependencies.R, since I've tried to minimize changes to the existing functions and files.  

This update recursively parses description files, downloading them if necessary, and sorts both dependencies and locked packages in the same list.  Dependencies are installed with locked packages in the same manner with the same function call, though dependencies will use the latest available version, only updating if packages' description file has updated it's version requirement beyond the one currently available in /.R/lockbox/  

This recursively parsed dependency information is NOT stored anywhere, but the parsing is done each time.  With a big lockfile it tends to add less than 2 seconds to startup, and certainly much less than the overall loading time.  Of course, when the packages have to be downloaded, this parsing can take longer, but still much less time than installation time.  There is some repetition of downloading here, so if we are unhappy with loading time on new package installations we can always add some additional logic to save paths to downloaded files

Currently, in the interest of leaving the original code as untouched as possible, I've left a little bit that could be dried up if combined with previous code, especially that in library.R.  The download_package.CRAN function shares some code with install_old_CRAN package, so that can be dried up if we want.

I'm accessing the Remotes portion of description files as well, but final dependencies are decided based on version comparison alone, since there is no real good way to compare two different forks of the same repo.  In addition, we currently only support github remotes, since that follows from the rest of lockbox.

Two private devtools functions are currently accessed in order to save space, remote_download.github_remote and github_remote, but we can import them in explicitly if we want to be safer.  At that point, however, I might just recommend dropping the devtools dependency entirely from lockbox, since we're just downloading all the same stuff in the same way.  Then we could also lock in devtools :)  I've also stolen the tools::package.dependencies function and updated it a bit so it doesn't break and gives us remotes back (this is the biggest function in dependencies.R....I hates it but don't want to touch it any more)

I've run devtools::document, so we've got a lot of extra files, but I've only altered by hand the existing files library.R, lockbox.R, and align.R (align is only changed to reduce the pesky_namespaces to lockbox's dependencies).

You'll find some logic that allows for NA versioning, which is acceptable for packages that have no version requirements.  Prior to installation they are updated with either the latest remote/CRAN version available (if we do not have the dependency or require a newer version) or the newest version in the lockbox. Some logic in lockbox.R is updated for this reason and also to distinguish, in super pretty crayon fashion, a dependency load or installation from a lockfile load or installation.

Finally, you should note that it's taken me more commits to get this right than it took for RobK to write the package in the first place.  Go me!

Commit:
NA version error handling

Commit:
wrong element

Commit:
Regex typo

Commit:
Installation cat error

Commit:
Clean up dependency flag logic

Commit:
Version specification for NA version

Commit:
fix na package logic

Commit:
Browsing again

Commit:
Fix dependency versioning behavior in install_old_CRAN...

Commit:
Tracking down dependency installation path error

Commit:
tracking down untar exception

Commit:
still tracking it down

Commit:
same again

Commit:
Fix tar version issue with CRAN dependency downloads

Commit:
Fix dependency package logic in install version check

Commit:
browser()

Commit:
not sure about this missing package bug

Commit:
Damn you munsell

Commit:
another failed library installation to .staging

Commit:
Browser

Commit:
Namespace issues

Commit:
Getting versions assigned prior to installation

Commit:
Small logic error fix

Commit:
Fix lazy logic

Commit:
Keep track of remote changes relative to master

Commit:
Wrong argument

Commit:
Wrong list

Commit:
It's a list object

Commit:
lapply not vapply, dummy

Commit:
Delete comment noise

Commit:
Version compare differing dependency repos.  Add more latest_version
logic

Commit:
Remove so old logic

Commit:
Add browser

Commit:
Errant argument

Commit:
swap errors

Commit:
browsin

Commit:
tracking down a duplicate

Commit:
tracking down

Commit:
still looking

Commit:
still hunting

Commit:
Add check for lockfile duplicates

Commit:
Fix github calls with no reference

Commit:
Github dependency fix when ref not present

Commit:
gmailr remote dependency issue

Commit:
Error in  Remotes dependencies install

Commit:
Same github installation woes

Commit:
command failed bug

Commit:
stub out install from devtools

Commit:
Stub better

Commit:
Figuring out stubbing

Commit:
Messing with devtools priority.  Probably a dead end

Commit:
Skipping devtools install because of dependency issues with R CMD
INSTALL

Commit:
Sigh.  Browsing in

Commit:
Unzipping manually because parsing can fail from zip with
install.packages apparently

Commit:
browse in

Commit:
syntax

Commit:
Add local install option for erring out install_github calls

Commit:
Looking good.  Keep minimal pesky namespaces

Commit:
Clean up mismatching code.  Remove library devtools call

Commit:
Extra argument

Commit:
add browser

Commit:
Modify latest_version  transfers to be coerced to a version correctly

Commit:
Pesky namespaces

Commit:
Don't install what's already there

Commit:
Only don't install pesky namespaces

Commit:
syntax

Commit:
Forgot argument

Commit:
Remove browser()

Commit:
Bad call to dependencies_from_description

Commit:
Parse dependencies from descriptions in lockbox path better

Commit:
Simplify some of the dependency addition logic

Commit:
Typo

Commit:
Missing argument

Commit:
Only read latest version from remote if local is unavailable

Commit:
Still too many remote calls

Commit:
Don't get remote info for base packages

Commit:
Still some queries on reload

Commit:
Version compare for local dependency packages

Commit:
Proper version comparison

Commit:
Remove browser

Commit:
Error in startup

Commit:
Remove pointless messages

Commit:
Missing latest version allowable

Commit:
Set backup github install to quiet

Commit:
Clean up comments/deprecated code to match master fork

Commit:
logic error.  Need to be able to have NA versions for dependencies

Commit:
Code placement

Commit:
Cleaning up code for potential review

Commit:
Use same roxygen version to minimize github diffs

Commit:
Keep version for dependencies as ref

Commit:
Delete unnecessary code

Commit:
Make sure we finalize dependencies in the right version format

Commit:
packages not printing

Commit:
Finalize as package_version

Commit:
Revert error handling on install_github...for now.

Commit:
Remove uneccesary line.  Cleanup comments

Commit:
Remove unnecessary function, clean up more comments

Commit:
fix logic error when dependencies can't be found

Commit:
Delete unnecessary lines

Commit:
Final version comparison update

Commit:
Deprecated argument

Commit:
Update documentation

Commit:
Remove some superfluous code

Commit:
Reorder to prettyify some code

Commit:
Run document again to clean up deleted functions

Commit:
remove redundant function

Commit:
Update version

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+#' along the way using the combine_dependencies function.
+get_dependencies_for_list <- function(master_list, lock, previously_parsed_deps) {
+  current_dependencies <- master_list
+  for (i in 1:length(master_list)) {
```
Use `seq_along`

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+
+#' Recursive function to take a list and lock and extract dependencies, sorting
+#' along the way using the combine_dependencies function.
+get_dependencies_for_list <- function(master_list, lock, previously_parsed_deps) {
```
Can you add Roxygen documentation and/or make it more clear what a `master_list` is?

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+#' Get dependencies for all elements in lock
+get_ordered_dependencies <- function(lock) {
+   cat(crayon::blue(paste("Retrieving dependency info...")))
+   deps <- get_dependencies_for_list(lock, lock, list())
```
`lock` twice? why?

davidboren replied:
Starts off the recursive chain.  Lock must be always available in it's original form, so that we know what we started out with.  The other option is to parse the list each time and look for !is_dependency_package, which should actually work fine, too.

davidboren replied:
I guess it's just me being paranoid about touching the original object

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+  for (i in 1:length(master_list)) {
+    package <- master_list[[i]]
+    previously_parsed_loc <- which_previously_parsed(
+        package, previously_parsed_deps)
```
Can you put more of things like this throughout on one line? It would make this block a lot more readable.

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+which_previously_parsed <- function(package, previously_parsed_deps) {
+  if (length(previously_parsed_deps) == 0) return(0L)
+  previously_parsed_names <- vapply(previously_parsed_deps
+    , function(p) p[["package"]]$name, character(1))
```
`p$package$name`?

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+    , function(p) p[["package"]]$name, character(1))
+  if (package$name %in% previously_parsed_names) {
+    sel <- which(previously_parsed_names == package$name)
+    subsel <- vapply(previously_parsed_deps[sel]
```
Can you not use `sel` as an intermediary variable?

Commit:
Add some documentation to main recursive file

Commit:
Use seq_along

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+    locked_package <- lock[[which(lock_names == package$name)[1]]]
+    if (!is.na(package$version) && package_version(as.character(package$version)) >
+      package_version(as.character(locked_package$version))) {
+        stop(paste0("Dependency: \'", package$name, ", Version: ", package$version
```
you can use `sQuote` instead of `\'`.  That should also let you use `paste` instead of `paste0`.

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+    if (!is.na(package$version) && package_version(as.character(package$version)) >
+      package_version(as.character(locked_package$version))) {
+        stop(paste0("Dependency: \'", package$name, ", Version: ", package$version
+          , " is required, but lockbox is locked at version: "
```
Can you say which package it is required by?

davidboren replied:
You know, I originally scrubbed that because of the way I was doing the package list combining, but, I think the way I have it now I can do it

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+      package_version(as.character(locked_package$version))) {
+        stop(paste0("Dependency: \'", package$name, ", Version: ", package$version
+          , " is required, but lockbox is locked at version: "
+          , as.character(locked_package$version)))
```
Maybe inform the user they should update their lockfile.

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+  } else {
+    package$is_dependency_package <- TRUE
+  }
+  if (!"remote" %in% names(package) || is.na(package$remote)) {
```
Does `names(package)` contain `NA`?  If not, you don't need the second part of this check.

If so, you could consider checking `%in% na.omit(names(package))`.

davidboren replied:
God.  It should not ever have NA.  I mean, I guess it's conceivable if the lockfile is reallly fucked up, but we explicitly add everything to it after that yml load

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+  }
+  package <- as.locked_package(package)
+  if (package$is_dependency_package && !package$name %in%
+    as.character(installed.packages(priority = "base")[,1])) {
```
What is priority for?

davidboren replied:
There are packages that are integrated into base now.  Trying to find them is a waste of time and a high potential for errors

peterhurford replied:
You should turn the search for installed packages into a reusable helper.

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+#' the parent package to the space after it's rightmost dependency found in list2.
+combine_dependencies <- function(list1, list2, current_parent) {
+  names1 <- vapply(list1, function(obj) obj$name, character(1))
+  names2 <- vapply(list2, function(obj) obj$name, character(1))
```
Two suggestions:

1.) Use `vapply(list1,`[[`, character(1), "name")`.

2.) DRY up by just processing as a list:

``` R
names <- lapply(list(list1, list2), function(lst) vapply(lst, `[[`, character(1), "name"))
```

and then use `names[[1]]` and `names[[2]]` instead of `names1` and `names2`.

davidboren replied:
Oh, crap, that's what was looking for.  Still was hoping for something even shorter. :D.  It's just a pluck.

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+
+  if (current_parent %in% names2 && any(names2 %in% names1)) {
+    init_parent_slot <- which(names2 == current_parent)
+    final_parent_slot <- max(which(names2 %in% names1))
```
Can you make these variable names more clear?

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+    final_parent_slot <- max(which(names2 %in% names1))
+    if (final_parent_slot > init_parent_slot) {
+      sel1 <- seq_along(names2) != init_parent_slot & seq_along(names2) <= final_parent_slot
+      sel2 <- seq_along(names2) > final_parent_slot
```
Can you make these variable names more clear? What's going on here?

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+
+  names(list1) <- names1
+  names(list2) <- names2
+  keep1 <- !names1 %in% names2
```
You can just use this down the road, no need for intermediary assignment

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+  c(list1[keep1], list2)
+}
+
+#' Swap all package information when side1 is greater than side2
```
what is `side1` and `side2`? Can you elaborate?

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+        obj2 <- list2[[n]]
+        if (obj1$is_dependency_package && obj2$is_dependency_package) {
+          if (is.na(obj1$version) && is.na(obj2$version)) {
+            if (obj1$remote != obj2$remote) {
```
Four nested `if` blocks usually means it's refactoring time.

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+              package_version(obj1$latest_version) > package_version(obj2$latest_version)
+            } else {
+              FALSE
+            }
```
You can do

``` R
obj1$remote != obj2$remote || package_version(obj1$latest_version) > package_version(obj2$latest_version)
```

Generally if an `if` block is being used to return a logical, you should instead just use whatever conditional is in the `if` block.

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+
+#' Swap all package information when side1 is greater than side2
+swap_versions <- function(names1, names2, list1, list2) {
+  swap_version2for1 <- vapply(
```
Can you give a clearer variable name for what this function does?

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+        }
+      } else{
+        FALSE
+      }}
```
e.g., this whole thing is:

``` R
compare_on_remotes <- function(obj1, obj2) {
  (obj1$remote != obj2$remote) &&
    (package_version(obj1$latest_version) > package_version(obj2$latest_version))
}
compare_on_versions <- function(obj1, obj2) {
  package_version(obj1$version) > package_version(obj2$version)
}
swap_version2for1 <- vapply(names1, function(n) {
  if (!(n %in% names2)) { return(FALSE) }
  obj1 <- list1[[n]]
  obj2 <- list2[[n]]
  not_dependency_package <- !obj1$is_dependency_package || !obj2$is_dependency_package
  not_dependency_package && !is.na(obj1$version) &&
    if (is.na(obj2$version)) {
      compare_on_remotes(obj1, obj2))
    } else {
      compare_on_versions(obj1, obj2)
    }
}, logical(1))
```

...though that involved a lot of transformations, so you should double check.

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+        FALSE
+      }}
+    , logical(1))
+  swap_versions1  <- names1 %in% names2 & swap_version2for1
```
This is just the same as `swap_version2for1`, as I understand it.

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+  swap_versions1  <- names1 %in% names2 & swap_version2for1
+  swap_versions2 <- vapply(names1[swap_versions1]
+    , function(n) which(names2 == n)
+    , integer(1))
```
How does this work? What is it supposed to be doing? I don't understand.

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+      package_version(as.character(locked_package$version)) <
+      package_version(as.character(locked_package$latest_version_in_lockbox)))) {
+      locked_package$version <- locked_package$latest_version_in_lockbox %||% NA
+  }
```
Some indermediary assignment would actually be eludicating here, to explain what each conditional is testing.

Maybe they could even become functions.

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+#' download the accurate remote DESCRIPTION file.
+get_remote_dependencies.CRAN <- function(package) {
+  if (package$is_dependency_package) {
+    package$version <- NA
```
Why is that?

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+  }
+  filepath <- download_package(package)
+  split_fp <- strsplit(filepath, "/")[[1]]
+  dirpath <- paste(split_fp[-length(split_fp)], collapse = "/")
```
R is one step ahead of you:

``` R
dirpath <- dirname(filepath)
```

peterhurford replied:
Also, for what it's worth, use `head(split_fp, -1)` instead of `split_fp[-length(split_fp)]`.

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+  dirpath <- paste(split_fp[-length(split_fp)], collapse = "/")
+  file_list <- untar(filepath, list = TRUE)
+  description_name <- file_list[grepl(paste0("^[^/]+"
+    ,"/DESCRIPTION$"), file_list)]
```
I think `grepl("DESCRIPTION", file_list, fixed = TRUE)` would also work and is safe enough?

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+#' Download the accurate remote DESCRIPTION file for a github repo.
+get_remote_dependencies.github <- function(package) {
+  dcf <- download_description_github(package)
+  dependencies_from_description(package, dcf)
```
No need for the intermediary assignment of `dcf`.

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+  description_path <- paste0(dirpath, "/", description_name)
+  dcf <- read.dcf(file = description_path)
+  unlink(filepath)
+  dependencies_from_description(package, dcf)
```
No need for the intermediary assignment of `dcf`.

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+    , subdir
+    ,"/DESCRIPTION$"), file_list$Name)]
+  file_con <- unz(filepath, description_name)
+  dcf <- read.dcf(file = file_con)
```
No need for the intermediary assignment of `dcf`.

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+  }
+  description_name <- file_list$Name[grepl(paste0("^[^/]+"
+    , subdir
+    ,"/DESCRIPTION$"), file_list$Name)]
```
Instead of:

``` R
  subdir <- ""
  if (!is.null(package$subdir)){
    subdir <- paste0("/",package$subdir)
  }
  description_name <- file_list$Name[grepl(paste0("^[^/]+"
    , subdir
    ,"/DESCRIPTION$"), file_list$Name)]
```

I think you can just do:

``` R
description_name <- grep(file.path(package$subdir, "DESCRIPTION"),
  file_list$Name, fixed = TRUE, value = TRUE)
```

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+}
+
+version_from_remote <- function(package) {
+  dcf <- download_description_github(package)
```
No need for the intermediary assignment of `dcf`.

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+#' Parse dependencies from description
+dependencies_from_description <- function(package, dcf) {
+  dependency_levels <- c("Depends", "Imports", "Remotes")
+  dependency_levels %in% colnames(dcf)
```
It looks like this condition isn't being used for anything?

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+}
+
+#' Parse dependencies from description
+dependencies_from_description <- function(package, dcf) {
```
Can you clean up this function? It's very hard to read and understand.

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+    filenames <- RCurl::getURL(url, ftp.use.epsv = FALSE, dirlistonly = TRUE)
+    filenames <- paste(url, strsplit(filenames, "\r*\n")[[1]], sep = "")
+    filenames <- filenames[grepl(paste0(name, "_[0-9\\.\\-]+\\.tar\\.gz"),filenames)]
+    expr <- gregexpr(paste0(name, "_([0-9\\.\\-]+)\\.tar\\.gz"), filenames)
```
What's the difference between `gregexpr` and `grep`?

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+  ###Some packages are available in archive only
+  if (is.na(remote_version)) {
+    archive_addition <- paste0("Archive/", name, "/")
+    url <- paste0(repo, "/src/contrib/", archive_addition)
```
Where is `repo` defined?

peterhurford replied:
Maybe

``` R
file.path(repo, "src", "contrib", "Archive", name)
```

?

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+  if (is.na(remote_version)) {
+    archive_addition <- paste0("Archive/", name, "/")
+    url <- paste0(repo, "/src/contrib/", archive_addition)
+    filenames <- RCurl::getURL(url, ftp.use.epsv = FALSE, dirlistonly = TRUE)
```
Maybe `httr::GET` instead?

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+    url <- paste0(repo, "/src/contrib/", archive_addition)
+    filenames <- RCurl::getURL(url, ftp.use.epsv = FALSE, dirlistonly = TRUE)
+    filenames <- paste(url, strsplit(filenames, "\r*\n")[[1]], sep = "")
+    filenames <- filenames[grepl(paste0(name, "_[0-9\\.\\-]+\\.tar\\.gz"),filenames)]
```
You should use `Filter` to narrow down a list instead of indexing

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+max_package_version <- function(versions) {
+  if (length(versions) == 0) return(NULL)
+  formatted_versions <- package_version(as.character(versions))
+  versions[which(formatted_versions == max(formatted_versions))[1]]
```
Use `Filter`?

robertzk replied:
Another way is

``` r
versions[Position(function(x) x == max(formatted_versions), formatted_versions)]
```

as in

``` r
> versions <- c("0.4.3", "0.3.3", "0.4.4", "0.4.1")
> formatted_versions <- package_version(as.character(versions))
> versions[Position(function(x) x == max(formatted_versions), formatted_versions)]
[1] "0.4.4"
```

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+replace_with_lock <- function(package, lock) {
+  lock_names <- vapply(lock, function(l) l$name, character(1))
+  if (package$name %in% lock_names) {
+    locked_package <- lock[[which(lock_names == package$name)[1]]]
```
`Filter`

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+    filenames <- filenames[grepl(paste0(name, "_[0-9\\.\\-]+\\.tar\\.gz"),filenames)]
+    expr <- gregexpr(paste0(name, "_([0-9\\.\\-]+)\\.tar\\.gz"), filenames)
+    filenames <- vapply(regmatches(filenames, expr), function(match) match[1], character(1))
+    archived_version <- gsub("(.*_)([0-9\\.\\-]+)(\\.tar\\.gz)", "\\2", filenames[length(filenames)])
```
Now I have no idea what is going on. Can you make it more clear?

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+  pkg_tarball <- tempfile(fileext = ".tar.gz")
+  out <- suppressWarnings(tryCatch(
+    download.file(url = from, destfile = pkg_tarball, quiet = notTRUE(getOption('lockbox.verbose')))
+    , error = function(e) e))
```
Just use `try` instead of `tryCatch` if you're not going to do anything with the error.

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+  }
+  if (nzchar(token <- Sys.getenv("GITHUB_PAT"))) {
+    arguments$auth_token <- token
+  }
```
Don't think you need the `if` protection here; I think it's okay for `arguments$auth_token` to be `""`.

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+        z <- unlist(strsplit(z, ",", fixed = TRUE))
+        z <- sub("^[[:space:]]*(.*)", "\\1", z)
+        z <- sub("(.*)[[:space:]]*$", "\\1", z)
+        pat <- "^([^\\([:space:]]+)[[:space:]]*\\(([^\\)]+)\\).*"
```
Oh my

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,510 @@
+                x1 <- format(x1, justify = "right")
+                x2 <- paste(x1[4:6], collapse = ".")
+                x1 <- paste(x1[1L:3], collapse = ".")
+                comptext <- paste0("'", x1, "' ", op, " '",
```
Haha this is all impossible to understand. ;)

peterhurford commented on R/lockbox.R:
```diff
@@ -122,3 +143,11 @@ disallow_special_packages <- function(lock) {
+    stop(paste0("The following packages are duplicated in your lockfile: "
+      , paste(locked_names[duplicated(locked_names)], collapse = ", ")))
+  }
+}
```
Nice :+1:

peterhurford commented on R/rebuild.R:
```diff
@@ -47,10 +47,8 @@ native_namespaces <-
-    "testthatsomemore", "stringr", "digest", "lubridate", "memoise",
-    "plyr", "magrittr", "devtools", "Rcpp", "roxygen")
+pesky_namespaces <- c("lockbox", "devtools", "yaml", "digest"
+  , "testthatsomemore" , "utils", "httr", "RCurl")
```
I think you can kill `RCurl` from here

peterhurford replied:
and `testthatsomemore`

peterhurford commented:
This is a lot of work -- congrats! :+1: Don't let my trivial comments detract from your accomplishment. ;)

Commit:
Peters suggested edits

Commit:
Use a tryCatch to catch installation errors, then re-run with verbose =
TRUE to get the full stacktrace if we did receive an error

Commit:
Remove leftover code

Commit:
Finish refactoring combine_dependencies

Commit:
Missing bracket

Commit:
Clean up some description file code

Commit:
Get rid of testthatsomemore dependency.  Investigate removing RCurl
dependency

Commit:
Browser for RCurl stuff

Commit:
Remove RCurl-dependent logic.  Move testhat to Imports

Commit:
Move testthat to Imports

Commit:
Manually import all devtools functions that we use

Commit:
Remove devtools

Commit:
Fix typo.  Remove some pesky namespaces

Commit:
Run document

Commit:
Replace testhatsomemore with testthat

Commit:
Run document

Commit:
Browser for remote failure

Commit:
Add additional stolen functions

Commit:
Split out devtools-derived functions

Commit:
Rearrange functions a bit

Commit:
Documentation

Commit:
More documentation

Commit:
More functions

Commit:
One, hopefully last, devtools function

Commit:
Remove browser

Commit:
Fix typo

Commit:
Put cran logic back in

Commit:
Tracking down recent update bug

Commit:
browser into cran

Commit:
Remove duplicate package downloading for consistency

Commit:
Document

Commit:
Fixing some typos in dependency returning

Commit:
Found version error.  Corrected

Commit:
update description

Commit:
add browser

Commit:
Browser

Commit:
show package name

Commit:
Set version from cran description because of shitting management

Commit:
Add additional latest version logic

Commit:
Check na

Commit:
fix latest version assignment order

Commit:
Browser into scales

Commit:
fix cran.  Remove extra crap

Commit:
Remove browser

Commit:
Need to be able to grab locked in versions that aren't archived

Commit:
If not dependency then no NA version

Commit:
Some explanation

Commit:
Browser one more time

Commit:
Improve link logic

Commit:
Unlink logic

Commit:
No devtools

Commit:
No browser

Commit:
Run document.  Add browser

robertzk commented:
Standby. This is going to require quite some review.

robertzk commented:
I actually love the idea of pulling in `remote_download.github_remote` and `github_remote` and getting rid of the devtools dependency. Can we do that?

Commit:
Fix arguments for S3 method

Commit:
Remove unnecessary line

Commit:
Need recursive unlinking

Commit:
Add chmod to config and cleanup files

Commit:
Errant temp statement removal

Commit:
Fix package object typo

Commit:
Add browser

Commit:
Remove browser()

Commit:
Do not load dependencies

Commit:
Stop on no dependency resolution

Commit:
Combine installation error logic

davidboren commented:
@robertzk Okay, took a little work but devtools is out of the picture as of the latest commit.  The borrowed functions are in the devtools_functions.R file and take up about 150 lines.  So our dependencies, not including the Suggests field, are down to httr, yaml, digest, utils, and crayon.  Installation from scratch is back up and running on my personal computer, but I'll have to wait until tommorow to test it on the ec2.  Because we aren't installing with devtools, I was able to remove the testthatsomemore dependency as well, since we can just pass in our library instead of playing make-believe with .libPaths.

Still need to clean up some of the description parsing stuff.  Too many gsubs in a few places.

Commit:
Modify installation logic a bit to only print installation line once,
to avoid confusing people with the second, verbose installation attempt

Commit:
Run document again

Commit:
Cleaning up a bit of code

Commit:
fix gsub typo

Commit:
merge upstream in

Commit:
Comment

Commit:
Unlink filepath correctly

Commit:
Styling

Commit:
Add try

Commit:
Comments

Commit:
Better list-storage of parsed dependencies

Commit:
Attach remote as class

Commit:
fix dependency list key from version

Commit:
Strip duplicate dependencies

Commit:
Try reversing the build

davidboren commented:
Okay, so I'm done with the initial refactoring and testing.  I've taken care of a ton of edge-cases without bloating the code too much.  Lockfiles build great from scratch.  Updating packages seems to be seamless.  We now download the packages only once, when we parse their dependencies, which saves us a little bit of time, but mostly ensures consistency between the packages used for parsing and those used for installation, since doing it twice with different methodologies was a recipe for disaster with some of CRAN's peculiarities..

kirillseva commented on DESCRIPTION:
```diff
@@ -2,26 +2,25 @@ Package: lockbox
 Title: lockbox (http://github.com/robertzk/lockbox)
 Description: Bundler-style dependency management for R.
-Version: 0.1.12
+Version: 0.1.13
```
0.2?
1.0?

peterhurford replied:
:+1:

I think it's 0.2 per [semver](http://semver.org/) since the change is backwards compatible.  Maybe 0.2.9000 until we know it's stable?

I think it will become 1.0 if/when we try to get it on CRAN. That's what Hadley seems to do, anyway.

kirillseva commented on DESCRIPTION:
```diff
@@ -2,26 +2,25 @@ Package: lockbox
-    testthatsomemore,
-    devtools,
-    utils
+    utils,
```
u da boss

robertzk replied:
Shouldn't be necessary to import utils?

kirillseva commented on NAMESPACE:
```diff
@@ -9,8 +9,6 @@ export(packageVersion)
-import(testthatsomemore)
 import(utils)
 import(yaml)
-importFrom(devtools,install_github)
```
:pray:

kirillseva commented on R/dependecies.R:
```diff
@@ -0,0 +1,593 @@
+#' Get dependencies for all elements in lock
+get_ordered_dependencies <- function(lock) {
+   cat(crayon::blue(paste("Retrieving dependency info...")))
```
paste?

kirillseva replied:
`cat(crayon::blue("Retrieving dependency info..."), "\n")` :facepunch:

kirillseva commented on R/dependecies.R:
```diff
@@ -0,0 +1,593 @@
+   cat(crayon::blue(paste("Retrieving dependency info...")))
+   names(lock) <- vapply(lock, `[[`, character(1), "name")
+   deps <- get_dependencies_for_list(lock, lock, list())
+   cat("\n")
```
huh, just return `get_dependencies_for_list(lock, lock, list())`

kirillseva commented on R/dependecies.R:
```diff
@@ -0,0 +1,593 @@
+    } else {
+      single_package_dependencies <-
+        previously_parsed_deps[[package$name]][[
+          paste(package$version)]]$dependencies
```
this looks like an awkward line break to me, but I see why you did that

davidboren replied:
I've been taking the column limit very seriously lately...

peterhurford replied:
You don't have to take it _that_ seriously :+1:

peterhurford replied:
Ironically, a lot of this code could be much easier to read if more things were on one line.

kirillseva commented on R/dependecies.R:
```diff
@@ -0,0 +1,593 @@
+
+      # Our package comes out of dependency search with a download_path
+      # attached.  We will use this for installation (assuming we keep this
+      # version of the package.
```
please use double pound sign for comments like this `##` so that it integrates with [rocco](https://github.com/robertzk/rocco)

kirillseva commented on R/dependecies.R:
```diff
@@ -0,0 +1,593 @@
+  }
+  filepath <- download_package(package)
+
+  sep <- .Platform$file.sep
```
wow

kirillseva commented on R/dependecies.R:
```diff
@@ -0,0 +1,593 @@
+
+#' Get the dependencies for a given package
+get_remote_dependencies <- function(package) {
+  UseMethod("get_remote_dependencies")
```
cool stuff

kirillseva commented on R/dependecies.R:
```diff
@@ -0,0 +1,593 @@
+  }
+}
+
+
```
![](http://www.reactiongifs.com/wp-content/uploads/2013/05/didnt-read.gif)

600 lines in one file T_T

davidboren replied:
I should probably put that last guy in it's own file.  impenetrable_regex_shit.R

kirillseva replied:
`never_open_this_file.R`

kirillseva commented:
if it works you're a wizard :stars: :fire:

Commit:
Split into multiple files vix comments to be rocco compatible

Commit:
Fix local package installation stuff.  Stop on unsupported Remotes

Commit:
typo

Commit:
Run document

davidboren commented:
Okay, just split into a couple additional files for clarity and I think I've addressed most of your comments to satisfaction.  Thanks for the notes.  Just locked in a local package with success so the logic is good on that front as well.  Also ran document again and had two successful installs and test_project()'s from scratch.

Commit:
Comments

Commit:
Up version

davidboren commented:
@robertzk any thoughts?  We can always use my fork for some of our more involved testing if you aren't comfortable merging this in at the moment.

robertzk commented on DESCRIPTION:
```diff
@@ -2,26 +2,25 @@ Package: lockbox
 Title: lockbox (http://github.com/robertzk/lockbox)
 Description: Bundler-style dependency management for R.
-Version: 0.1.12
+Version: 0.2.9000
```
Can you use 0.2.0.9000 please? Development version should be 4th index.

peterhurford replied:
:+1:

robertzk commented on DESCRIPTION:
```diff
@@ -2,26 +2,25 @@ Package: lockbox
     R (>= 3.0.1),
-    httr,
-    RCurl
+    httr
```
Can you move this to imports instead of depends?

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+    if(!package$name %in% names(previously_parsed_deps)) {
+      previously_parsed_deps[[package$name]] <- list()
+    }
+    if (!paste(package$version) %in% 
```
Why a wrapper `paste` here?

robertzk replied:
This whole `if` statement can be one line.

davidboren replied:
paste comes from my using the NA value for version to mean "latest-available"
I could just use "latest-available," since that wouldn't collide with any version.  It would still, however, not be convertible to package_version...

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+    if (!paste(package$version) %in% 
+      names(previously_parsed_deps[[package$name]])) {
+      dependency_output <- get_dependencies(
+       structure(package
```
I think it's fine to put this on one line.

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+
+      ## Store the dependencies and corresponding package object in our humongous
+      ## previously_parsed_deps object using version as the second key level
+      previously_parsed_deps[[package$name]][[paste(package$version)]] <-
```
Ah, I see. You can use `as.character` instead of `paste` for a more natural interpretation of converting the `package_version` S3 object to a stirng.

davidboren replied:
Yeah, but the NA versioning was giving me trouble :) Because I allow NA to represent no version requirements.  Could be something else, I guess.  Yeah, also because of that dependency package versions aren't converted to package_version objects until the very end (inside the reset_to_latest_version function).

peterhurford replied:
`paste(NA)` gives the literal string `"NA"` which no longer means `NA` though. Sounds dangerous?

davidboren replied:
Yeah, I it needs to be a string to act as a key.  If we wanted to we could use some other string to represent "any version available," but then we still couldn't convert to package_version, right?  The paste is there to produce the key.  We never convert the actual package$version entry.

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+      ## previously_parsed_deps object using version as the second key level
+      previously_parsed_deps[[package$name]][[paste(package$version)]] <-
+        list(package = package, dependencies = single_package_dependencies)
+    } else {
```
I would swap these two blocks -- `if (as.character(package_version) %in%  names(previously_parsed_deps[[package$name]]))` seems more natural than its negation.

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+      , package$name)
+  }
+
+  if (identical(master_list, current_dependencies)) return(master_list)
```
Maybe `if/else` here to make the dichotomy clearer?

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+  if (identical(master_list, current_dependencies)) return(master_list)
+
+  ## Run through the entire list again, because we have a bunch of new packages
+  get_dependencies_for_list(current_dependencies, lock, previously_parsed_deps)
```
Prefer `Recall` over explicit recursion naming.

davidboren replied:
Good point.

kirillseva replied:
:+1: to recall

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+#' @param lock list.  Original list of locked packages
+#' @param previously_parsed_deps.  List of packages and their dependencies that
+#'   we have already parsed out of their respective description files
+get_dependencies_for_list <- function(master_list, lock, previously_parsed_deps) {
```
Let me summarize to make sure I understand correctly. If the tree of dependencies runs to depth 5, this will recursively call itself 5 times until all the dependencies have been exhausted into the `master_list`, at which point it will have no further updates and the `identical` clause below terminates the entire chain?

davidboren replied:
Yep.  That's about right.  Because we use version comparison to decide on the "winners" of deciding between dependencies, the dependency list we wind up with will be stable and unmodified in the last pass.  It will also be much faster since once a package is parsed for it's dependencies they are saved into the previously_parsed_dependencies list.

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+#' with the locked version if it does appear there.  Also throw
+#' an error if we require a dependency version greater than that specified by
+#' the lockfile.
+replace_with_lock <- function(package, lock) {
```
`check_doc` will complain if you don't add the `params` but use literate documentation. (You can also turn `#'` into `#` if you want plain docs)

peterhurford replied:
Gotta pass `R CMD CHECK`...

kirillseva replied:
or ##

davidboren replied:
Good to know!

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+replace_with_lock <- function(package, lock) {
+  lock_names <- vapply(lock, function(l) l$name, character(1))
+  if (package$name %in% lock_names) {
+    locked_package <- Filter(function(l) l$name == package$name, lock)[[1]]
```
`Filter(...)[[1]]` is equivalent to `Find(...)`

peterhurford replied:
Did not know

davidboren replied:
Yeah.  That's a new one to me.  good FIND

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+
+#' Recursive function to take a list and lock and extract dependencies, sorting
+#' along the way using the combine_dependencies function.
+#' @param master_list list.  Packages to go through and parse dependencies from
```
Can you call this something else, like `package_list`?

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+  lock_names <- vapply(lock, function(l) l$name, character(1))
+  if (package$name %in% lock_names) {
+    locked_package <- Filter(function(l) l$name == package$name, lock)[[1]]
+    if (!is.na(package$version) && package_version(as.character(package$version)) >
```
Isn't `package$version` already a `package_version`?

davidboren replied:
Not for dependency packages, due to to package_version's incompatibility with NA versions.  Even if, instead of NA, we used some other version to represent "latest_version_available" we would still have problems with either conversion to package_version (if it doesn't recognize the format) or collision with actual versions (if it does)

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+          , " is required by package ", package$parent_package
+          , ", but lockbox is locked at version: "
+          , as.character(locked_package$version)
+          , ". Please update your lockfile accordingly"))
```
Nice!

kirillseva replied:
maybe add manual crayon? :D

robertzk replied:
Yeah - manual crayon is trivial -- just `print(crayon::foo("bar"))` and copy paste the results.

peterhurford replied:
Yeah manual crayon should be pretty easy because we don't use the full extent of crayon in this package.  Just need three functions for the three different colors we use, perhaps in `colors.R`.

davidboren replied:
Hell yeah.  One more dependency down.

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+
+#' Combine two lists of dependencies via version comparisons.  Keep packages
+#' found in list1 on the left side of the entirety of list2, while moving
+#' the parent package to the space after it's rightmost dependency found in list2.
```
Can you leave an example here? Very hard to decipher without knowing what sample inputs look like.

robertzk replied:
Can you also explain what `current_parent` is?

robertzk replied:
Ok I think I understand now. So you're saying that after combining, the final order of installation will be left-to-right, so you are "bubbling" the packages with dependencies to their right so that they occur later in the final sequence? I'm thinking about whether there is a cleaner, more understandable way. One way that might not be feasible given R's lack of big integer support is:
1. Tag each dependency-less package with an increasing integer sequence.
2. Tag each 1-dependency package with the sequence, continually increasing.
   etc.

Until we've run out of packages. Then we just `order()` by the tags! Is that doable?

robertzk replied:
Ah, nevermind, I see. You also have to deal with issues where there is the same package with a different version in each list.

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+  names <- lapply(list(list1, list2), function(lst) vapply(lst, `[[`, character(1), "name"))
+
+  ## Certain packages are no longer on cran but incorporated into R Core
+  core_pkgs <- as.character(installed.packages(priority = "base")[,1])
```
Seeing this four times seems repetitive but I can't think of a cleaner way!

peterhurford commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+  list1 <- list1[!names[[1]] %in% core_pkgs]
+  names[[1]] <- names[[1]][!names[[1]] %in% core_pkgs]
+  list2 <- list2[!names[[2]] %in% core_pkgs]
+  names[[2]] <- names[[2]][!names[[2]] %in% core_pkgs]
```
You can avoid this duplication by working with `lists <- list(list1, list2)` instead.

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+  if (length(list1) == 0) return(list2)
+  if (length(list2) == 0) return(list1)
+
+  names(list1) <- names[[1]]
```
Why is this necessary? You already performed dual stripping operations above.

davidboren replied:
Hmmm, I've referenced priority twice, but once is just a check without stripping. Should be stripped right before the replace_with_lock lapply is run instead.  On it.

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+  ## Find the rightmost dependency of the current package and move our current
+  ## package to the immediate right of that spot.  This preserves previous
+  ## sorting order while ensuring that the current package will be installed
+  ## after all its dependencies
```
So clearly this function is non-commutative. What is the criterion for putting a dependency list as list2 instead of list1?

davidboren replied:
You're right.  That'd be cray.  I've updated the logic to reflect as much.

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+  ## after all its dependencies
+  if (current_parent %in% names[[2]] && any(names[[2]] %in% names[[1]])) {
+    init_parent_slot <- which(names[[2]] == current_parent)
+    final_parent_slot <- max(which(names[[2]] %in% names[[1]]))
```
`which.max` :)

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+swap_packages <- function(names1, names2, list1, list2) {
+  ## Swap packages when the package in list1 is more recent (> version) than
+  ## it's corresponding package in list2.  Keep the list 2 element if it is a
+  ## locked package, as well.  If their remotes are not identical and neither
```
What if the remotes are identical and `list2`'s version is greater? Then don't swap...?

davidboren replied:
You know, I don't think I need the non-identical check.  Version should win out regardless.  Good catch.

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+      obj1 <- list1[[index]]
+      if (!is.null(obj1$download_path) && !is.null(obj2$download_path) &&
+        obj2$download_path != obj1$download_path) {
+        unlink(obj2$download_path, TRUE, TRUE)
```
Shouldn't the directory with downloaded files be unlinked anyway after the `lockbox` call? This may be redundant.

davidboren replied:
Downloads and extractions are now done in a directory given by the function call lockbox_download_dir()

elainexmas commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+#' Get dependencies for all elements in lock
```
Can you fix the spelling on the filename, 'dependencies.R'?  Thanks.

davidboren replied:
Good catch.  done.  :)

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+      }})
+
+  ## Swap a package from list1 into list2
+  list2_swap <- vapply(names1[swap_package2for1]
```
You re-invented `match` :)

``` r
names1 <- c("foo", "bar", "baz", "qux")
names2 <- c("baz", "foo", "qux", "bar")
swap_package2for1 <- c(1, 3)
list2_swap <- vapply(names1[swap_package2for1], function(n) which(names2 == n), integer(1))
alternative <- match(names1[swap_package2for1], names2)
stopifnot(identical(unname(list2_swap), alternative))
```

davidboren replied:
I tend to do that :/
Always coding before googling...

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+}
+
+strip_available_dependencies <- function(dependencies) {
+  dependencies <- lapply(dependencies
```
Why not just

``` r
strip_available_dependencies <- function(dependencies) {
  dependencies[!vapply(dependencies, `[[`, "name", character(1)) %in% pesky_namespaces]
}
```

davidboren replied:
we can do that

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+ dependencies[!vapply(dependencies, is.null, logical(1))]
+}
+
+#' We don't trust package authors to only put a package in once
```
Never trust anyone!

robertzk replied:
Everybody lies.

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+  UseMethod("get_remote_dependencies")
+}
+
+#' If a package is local we just read from the directory given
```
You need to stop using literate documentation if you're not actually going to comment the params and return value :)

robertzk replied:
This is going to give so many doc errors

davidboren replied:
not anymore!

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+
+#' If a package is local we just read from the directory given
+get_remote_dependencies.local <- function(package) {
+  description_name <- paste0(package$dir, .Platform$file.sep, "DESCRIPTION")
```
Always prefer `file.path(package$dir, "DESCRIPTION")`

davidboren replied:
done

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+  split_fp <- strsplit(filepath,sep)[[1]]
+  dirpath <- dirname(filepath)
+  file_list <- untar(filepath, list = TRUE)
+  description_name <- file_list[grepl(paste0("^[^", sep, "]+", sep
```
Wait, why do we need that prefix? What's wrong with just `description_name = grep("^.?DESCRIPTION$", file_list, value = TRUE)`?

robertzk replied:
Surely no author is evil enough to include a `.DESCRIPTION` file or some nonsense :)

davidboren replied:
I ran into an evil bastard when I was stress-testing this...

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+  description_name <- file_list[grepl(paste0("^[^", sep, "]+", sep
+    ,"DESCRIPTION$"), file_list)]
+  output <- untar(filepath, description_name, exdir = dirpath)
+  description_path <- paste0(dirpath, sep, description_name)
```
`file.path` please

davidboren replied:
good catch

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+  package$version <- original_version
+  dcf <- read.dcf(file = description_path)
+  package$latest_version <- version_from_description(package$name, dcf)
+  unlink(description_path)
```
Whaaaat? You're leaving the downloaded directory but removing just the description file? This makes no sense to me.

davidboren replied:
that should have been dirpath.  Good catch

Commit:
Version, robk's comments.  Some styling changes

Commit:
Rename dependencies

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+  if (!is.null(package$subdir)){
+    subdir <- paste0(sep, package$subdir)
+  }
+  description_name <- file_list$Name[grepl(paste0("^[^", sep, "]+"
```
You've already included `sep` in `subdir` above and are using `+` matching. This means you'll be looking for things like `//foo/DESCRIPTION` -- how is that possibly going to match? In general, I would prefer something like:

``` r
file_list <- Filter(function(x) substring(x, 1, nchar(subdir)) == subdir, file_list)
description_name <- Find(function(x) basename(x) == "DESCRIPTION", file_list)
```

robertzk replied:
Unless you're saying `file_list` is prefixed by the `sep`? In which case use `2` instead of `1` in the `substring` above.

davidboren replied:
So these packages always unzip into some directory defined in the zip file, So I'm looking for non-file-separators followed by a separator ([^/]+/) to get the description file in the immediate directory above, unless there is the subdirectory, in which case it is two directories up.  It's almost guaranteed to be the first thing that matches DESCRIPTION at all in file_list$Name, but after seeing tons of weird cran and package behaviors I'm super paranoid and wanted to be very explicit.

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+
+version_from_remote <- function(package) {
+  output <- download_description_github(package)
+  unlink(package$download_path)
```
Always use `force = TRUE, recursive = TRUE`

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+
+version_from_description <- function(package_name, dcf) {
+  if (!"Version" %in% colnames(dcf)) return(NA)
+  as.character(dcf[1, which(colnames(dcf) == "Version")])
```
`which` is unnecessary here. Can you make this more natural?

``` r
version_from_description <- function(package_name, dcf) {
  if (is.element("Version", colnames(dcf))) {
    as.character(dcf[1, "Version"])
  } else {
    NA_character_
  }
}
```

davidboren replied:
Done

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+    , non_remote_dependencies[,1]), , drop = FALSE]
+
+  ## Parse the non-remote dependencies into a list of packages
+  if (identical(nrow(non_remote_dependencies),0L)){
```
`NROW` over `nrow`

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+  if (identical(nrow(non_remote_dependencies),0L)){
+    non_remote_list <- list()
+  } else {
+    non_remote_list <- lapply(seq_along(non_remote_dependencies[,1])
```
I prefer `[[1]]` here given how much I hate type dropping :)

robertzk replied:
I am so confused. How is this not just

``` r
non_remote_list <- list(name = non_remote_dependencies[, 1], version = non_remote_dependencies[, 3])
```

davidboren replied:
You know what, it's a friggin map:

``` r
non_remote_list <- Map(function(u,v) list(name = u, version = v)
  , as.character(non_remote_dependencies[,1]), as.character(non_remote_dependencies[,3]))
```

robertzk replied:
Yes. There you go

On Mon, Feb 29, 2016 at 5:18 PM, davidboren notifications@github.com
wrote:

> In R/dependecies.R
> https://github.com/robertzk/lockbox/pull/73#discussion_r54497329:
> 
> > +
> > -  ## We separate out non-remote dependencies from remote dependencies, because
> > -  ## they require different logic
> > -  non_remote_dependencies <- dependencies_parsed[rownames(dependencies_parsed) != "Remotes", ]
> > -  remote_dependencies <- dependencies_parsed[rownames(dependencies_parsed) == "Remotes", ]
> >   +
> > -  ## Remove the Depends entry that just corresponds to the R version
> > -  ## requirements
> > -  non_remote_dependencies <- non_remote_dependencies[!grepl("^[rR]$"
> > -    , non_remote_dependencies[,1]), , drop = FALSE]
> >   +
> > -  ## Parse the non-remote dependencies into a list of packages
> > -  if (identical(nrow(non_remote_dependencies),0L)){
> > -    non_remote_list <- list()
> > -  } else {
> > -    non_remote_list <- lapply(seq_along(non_remote_dependencies[,1])
> 
> You know what, it's a friggin map:
> 
> non_remote_list <- Map(function(u,v) list(name = u, version = v)
>   , as.character(non_remote_dependencies[i,1]), as.character(non_remote_dependencies[i,3]))
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/robertzk/lockbox/pull/73/files#r54497329.

## 

-Robert Krzyzanowski

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+      })
+  }
+
+  ## Parse the remote dependencies into a list of packages
```
NROW

Commit:
Remove crayon dependencies

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+  if (identical(nrow(remote_dependencies),0L)){
+    remote_list <- list()
+  } else {
+    matches_github <- grepl("github::"
```
`fixed = TRUE` if not using regex plz

davidboren replied:
nice

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+      stop(paste0("Package ", package$name, " from repo ", package$repo
+        , " has unsupported (non-github) remote dependencies"))
+    } else {
+      remote_list <- lapply(seq_along(remote_dependencies[,1])
```
Whaaaat? Why not just

```r
lapply(remote_dependencies[, 1], function(name) { 
...
```

robertzk replied:
Please turn the inner loop into a helper function that takes `name` and `version`, and then call

```
remote_list <- Map(extract_package_from_remote, remote_dependencies[, 1], remote_dependencies[, 3])
```

robertzk replied:
This is too long to put in another function! Your functions should be 10 lines of code max.

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+  if (length(remote_list) != 0) {
+    non_remote_names <- vapply(non_remote_list, function(pkg) pkg$name, character(1))
+    remote_names <- vapply(remote_list, function(pkg) pkg$name, character(1))
+    if (any(non_remote_names %in% remote_names)) {
```
Remove the `if`.

robertzk commented on R/dependecies.R:
```diff
@@ -0,0 +1,437 @@
+  }
+
+  ## Remote package names are duplicated in Depends, LinkingTo, and Imports entries
+  if (length(remote_list) != 0) {
```
Just remove this whole `if` block and turn it into

```
non_remote_list <- non_remote_list[!is.element(
  vapply(non_remote_list, `[[`, character(1), "name"), vapply(remote_list, `[[`, character(1), "name"))]
```

davidboren replied:
nice

robertzk commented on R/devtools_substitutes.R:
```diff
@@ -0,0 +1,156 @@
+}
+
+parse_git_repo <- function(path) {
+  username_rx <- "(?:([^/]+)/)?"
```
u cray haddawg

robertzk commented on R/library.R:
```diff
@@ -25,123 +25,86 @@ exists_in_lockbox <- function(locked_package) {
-    devtools::install(locked_package$dir,
-                      quiet = notTRUE(getOption("lockbox.verbose"))))
+
+  utils::install.packages(locked_package$dir, lib = libP, repos = NULL
```
Thanks

robertzk commented on R/library.R:
```diff
@@ -25,123 +25,86 @@ exists_in_lockbox <- function(locked_package) {
-      quiet = notTRUE(getOption('lockbox.verbose'))))
-  }
+install_package.CRAN <- function(locked_package, libP, quiet) {
+  filepath <- locked_package$download_path
```
Clever

robertzk commented on R/library.R:
```diff
@@ -25,123 +25,86 @@ exists_in_lockbox <- function(locked_package) {
+  ## while maintaining cross-platform support
+  extracted_dir <- gsub(file.path(parent_dir, "")
+    , "", unzip(filepath, exdir = parent_dir)[1])
+  extracted_dir <- file.path(parent_dir
```
`normalizePath(file.path(parent_dir, extracted_dir))` should be just fine

robertzk commented on R/library.R:
```diff
@@ -25,123 +25,86 @@ exists_in_lockbox <- function(locked_package) {
-#' @importFrom devtools install_github
-install_package.github <- function(locked_package) {
-  stopifnot(is.element("repo", names(locked_package)))
+  ## It's absolutely necessary that we make these executable
```
What if they don't exist...?

davidboren replied:
Apparently R doesn't care.  That's either a good thing or another example of R being flexible to the point of madness.

robertzk commented on R/library.R:
```diff
@@ -191,11 +154,102 @@ current_version.locked_package <- function(package) {
-description_file_for <- function(package_name) {
-  dcf_file <- file.path(libPath(), package_name, "DESCRIPTION")
+description_file_for <- function(package_name, libP) {
+  dcf_file <- file.path(libP, package_name, "DESCRIPTION")
```
Can you `s/libP/libPath` on all of these? I prefer explicit naming.

davidboren replied:
done

Commit:
browser for improving download_description_github logic

robertzk commented on R/library.R:
```diff
@@ -191,11 +154,102 @@ current_version.locked_package <- function(package) {
+  ## things have been updated on the information page for a package but others
+  ## have not.
+  if (package$is_dependency_package) {
+    pkg_tarball <- tempfile(fileext = ".tar.gz")
```
No no no. We should be using the lockbox path (`~/.R/lockbox`) `/download` so we can just unlink the whole damn thing after everything's done and over with.

davidboren replied:
I was doing something similar at one point, then decided to make use of temp :(
Will refactor

robertzk replied:
Thank you!

robertzk commented on R/library.R:
```diff
@@ -191,11 +154,102 @@ current_version.locked_package <- function(package) {
+  if (package_version(remote_version) == package_version(version)) {
+    version <- remote_version
+    archive_addition <- ""
+  } else{
```
Add a space

robertzk commented on R/library.R:
```diff
@@ -191,11 +154,102 @@ current_version.locked_package <- function(package) {
+  from <- paste0(repo, "/src/contrib/", archive_addition, name, "_", version
+    , ".tar.gz")
+
+  pkg_tarball <- tempfile(fileext = ".tar.gz")
```
Same

robertzk commented on R/library.R:
```diff
@@ -191,11 +154,102 @@ current_version.locked_package <- function(package) {
+
+#' Create remote in form devtools' remote_download  function likes.
+get_remote <- function(package) {
+  ref <- package$ref %||% package$version
```
Other way around!! `version` can override `ref`, not conversely.

robertzk commented on R/library.R:
```diff
@@ -191,11 +154,102 @@ current_version.locked_package <- function(package) {
+  if (is.null(ref)) {
+    arguments <- list(package$repo)
+  } else {
+    arguments <- list(
```
One line plz

robertzk commented on R/library.R:
```diff
@@ -191,11 +154,102 @@ current_version.locked_package <- function(package) {
+    arguments <- list(
+      paste(package$repo, ref, sep = "@"))
+  }
+  if (nzchar(token <- Sys.getenv("GITHUB_PAT"))) {
```
Explicit style plz

``` r
token <- Sys.getenv("GITHUB_PAT")
if (nzchar(token)) ...
```

davidboren replied:
Ahhh, dat legacy code from stealing crap :)

robertzk commented on R/library.R:
```diff
@@ -191,11 +154,102 @@ current_version.locked_package <- function(package) {
+  if (!is.null(package$subdir)) {
+    arguments$subdir <- package$subdir
+  }
+  remote <- do.call(github_remote, arguments)
```
rm `remote <-`

robertzk commented on R/lockbox.R:
```diff
@@ -40,22 +40,32 @@ lockbox.list <- function(lock, env) {
-    ## And re-build our search path.
-    rebuild(lock)
+    ## And re-build our search path. Do so in the reverse order of dependencies.
+    rebuild(rev(lock))
```
:100:

robertzk commented on R/lockbox.R:
```diff
@@ -68,22 +78,34 @@ lockbox.default <- function(obj) {
 as.locked_package <- function(list) {
   stopifnot(is.element("name", names(list)),
             is.element("version", names(list)))
 
```
Better is

``` r
list$is_dependency_package <- isTRUE(list$is_dependency_package %||% FALSE)
```

robertzk commented on R/lockbox.R:
```diff
@@ -68,22 +78,34 @@ lockbox.default <- function(obj) {
   }
 
-  # TODO: (RK) Support CRAN version dependencies.
+  ##  TODO: (RK) Support CRAN version dependencies.
```
@davidboren Is that addressed by this PR?

davidboren replied:
Why yes it is :)  Will destroy

robertzk commented on R/lockbox.R:
```diff
@@ -122,3 +144,11 @@ disallow_special_packages <- function(lock) {
   }
 }
+
+disallow_duplicate_packages <- function(lock) {
```
:100:

robertzk commented on R/lockbox.R:
```diff
@@ -122,3 +144,11 @@ disallow_special_packages <- function(lock) {
+  locked_names <- vapply(lock, function(p) p$name, character(1))
+  if (any(duplicated(locked_names))) {
+    stop(paste0("The following packages are duplicated in your lockfile: "
+      , paste(locked_names[duplicated(locked_names)], collapse = ", ")))
```
`unique()` around this so you don't print the names multiple times in case of 3+ duplications

robertzk commented on R/tools_substitutes.R:
```diff
@@ -0,0 +1,60 @@
+#' Borrowed from tools::package.dependencies and modified to be less breaky
```
Please detail the modifications.

robertzk commented on R/tools_substitutes.R:
```diff
@@ -0,0 +1,60 @@
+  for (k in 1L:nrow(x)) {
+    z <- x[k, colnames(x) %in% depLevel]
+    if (all(!is.na(z)) & all(z != "")) {
+        z <- unlist(strsplit(z, ",", fixed = TRUE))
```
Seriously R core? Seriously?

robertzk commented on R/tools_substitutes.R:
```diff
@@ -0,0 +1,60 @@
+          if (!is.na(deps[[k]]) && any(ok <- deps[[k]][, 1] ==
+              "R")) {
+              if (!is.na(deps[[k]][ok, 2]) && deps[[k]][ok,
+                2] %in% c("<=", ">=")) {
```
My eyes are bleeding R core team

robertzk commented:
Ok, so here's the deal. This is a formidable tour de force and I am happy to accept it into the realm of the lockbox legion. However, since this package will potentially be bundled with the open source Syberia release and disseminated to the wide world of the R bourgeoisie, it is imperative we perform the following before merging to this repository.

I would like a script (it can be a simple shell script, nothing too complicated) that
1. 1000 times does:
2. Sample 20 packages from all CRAN packages and add 5-10 robertzk/syberia packages
3. Create a lockfile in random order with those packages, set lockbox path to new directory.
4. Run `Rscript -e "lockbox::lockbox('generated_lockfile.yml');cat('Successfully loaded lockbox')" > some_logfile` and then check
   that `some_logfile` contains `Successfully loaded lockbox` (can't rely on system status as I'm sure it will falsely report 0 in some weird situations).

This will give us quickcheck-level of confidence that it works in a large variety of situations with all CRAN-approved R packages. If some packages end up being problematic we can find them independently and mark them as bad. 

I know it's a good bit of more work, but certainly not nearly as much as it took to get here in the first place.

After that we'll set the banquet and feast!

peterhurford commented:
Commits 250+ nice...

peterhurford commented:
@robertzk @davidboren Maybe make include an R script to automatically do https://github.com/robertzk/lockbox/pull/73#issuecomment-190432973 so it can be a reproducible test (given a seed)?

robertzk commented:
You want to recursively call R from R using `system` in a test!? 

Okay madmen. Okay.

Commit:
Robk suggestions

Commit:
More Robk suggestions

Commit:
Create download directory at appropriate time

Commit:
Rename typo

Commit:
Fix order in vapply statements

Commit:
Remove download dir correctly

Commit:
Tracking down bug introduced by refactoring

Commit:
Fix name mismatch

Commit:
Browser()

Commit:
Beautify some filepath stuff

Commit:
Run document

Commit:
Fix comment formatting

Commit:
document and pass documnet checks

Commit:
filepath finalizing

Commit:
Finish "final" debug

davidboren commented:
Okay.  I think I've gotten all of the comments and suggestions integrated.  We are building lockfiles from scratch with no problem.  I sleep now.

Tomorrow we can flesh out that awesome test idea.

Commit:
Add in ref logic for cran packages

Commit:
Edge-cases for CRAN and add ref logic

robertzk commented:
Excellent. Let's see what the results are and we can merge when we have good coverage over random lockfiles.

Commit:
Document. But more importantly Fix s3 bug where env is not passed into
generic-specific s3 methods

Commit:
Messaging for ci

Commit:
Better messaging for drone

Commit:
newline

Commit:
More debugging for drone

Commit:
Debugging

davidboren commented:
Decided to take a brake from debugging some of those random lockbox peculiarities and started making that wacky lockbox test.  Progress:

``` r
R-EXP> github_packages <- yaml::yaml.load_file(file.path(test_dir, "github_packages.yml"))$packages[1]
R-EXP> cran_packages <- get_random_cran_packages(2)
R-EXP> get_result(github_packages, cran_packages, test_dir, FALSE)
 [1] "\033[31menv is !packages"
 [2] "\033[39m\033[31mLock has packages: stagerunner, translateR, MatchLinReg"
 [3] "\033[39m\033[34mRetrieving dependency info...\033[39m\033[34m.\033[39m\033[34m.\033[39m\033[34m.\033[39m\033[34m.\033[39m\033[34m
.\033[39m\033[34m.\033[39m\033[34m.\033[39m\033[34m.\033[39m\033[34m.\033[39m\033[34m.\033[39m\033[34m.\033[39m\033[34m.\033[39m\033[34
m.\033[39m"
 [4] "Using dependency \033[34mstringi\033[39m 1.0.1 "
 [5] "Using dependency \033[34mmagrittr\033[39m 1.5 "
 [6] "Using dependency \033[34mcolorspace\033[39m 1.2.6 "
 [7] "Using dependency \033[34mdichromat\033[39m 2.0.0 "
 [8] "Using dependency \033[34mmunsell\033[39m 0.4.3 "
 [9] "Using dependency \033[34mlabeling\033[39m 0.3 "
[10] "Using dependency \033[34mstringr\033[39m 1.0.0 "
[11] "Using dependency \033[34mRcpp\033[39m 0.12.3 "
[12] "Using dependency \033[34mRColorBrewer\033[39m 1.1.2 "
[13] "Using dependency \033[34mplyr\033[39m 1.8.3 "
[14] "Using dependency \033[34mreshape2\033[39m 1.4.1 "
[15] "Using dependency \033[34mscales\033[39m 0.4.0 "                                                                                  
[16] "Using dependency \033[34mmemoise\033[39m 1.0.0 "
[17] "Using dependency \033[34mbitops\033[39m 1.0.6 "
[18] "Using dependency \033[34mMASS\033[39m 7.3.45 "
[19] "Using dependency \033[34mlattice\033[39m 0.20.33 "
[20] "Using dependency \033[34msurvival\033[39m 2.38.3 "
[21] "Using dependency \033[34mrpart\033[39m 4.1.10 "
[22] "Using dependency \033[34mgtable\033[39m 0.2.0 "
[23] "Using dependency \033[34mggplot2\033[39m 2.1.0 "
[24] "Using dependency \033[34mgridExtra\033[39m 2.2.1 "
[25] "Using dependency \033[34mRJSONIO\033[39m 1.3.0 "
[26] "Using dependency \033[34mRCurl\033[39m 1.95.4.8 "
[27] "Using dependency \033[34mcrayon\033[39m 1.3.1 "
[28] "Using dependency \033[34mR6\033[39m 2.1.2 "
[29] "Using \033[32mstagerunner\033[39m 0.5.2 "
[30] "Installing dependency \033[34mtau\033[39m from CRAN "
[31] "Installing dependency \033[34mslam\033[39m from CRAN "
[32] "Installing dependency \033[34mFormula\033[39m from CRAN "
[33] "Installing dependency \033[34mlatticeExtra\033[39m from CRAN "
[34] "Installing dependency \033[34mcluster\033[39m from CRAN "
[35] "Installing dependency \033[34mnnet\033[39m from CRAN "
[36] "Installing dependency \033[34macepack\033[39m from CRAN "
[37] "Installing dependency \033[34mforeign\033[39m from CRAN "
[38] "Installing dependency \033[34mHmisc\033[39m from CRAN "
[39] "Installing dependency \033[34mMatching\033[39m from CRAN "
[40] "Installing dependency \033[34mtextcat\033[39m from CRAN "
[41] "Installing \033[32mtranslateR\033[39m 1.0 from CRAN "
[41] "Installing \033[32mtranslateR\033[39m 1.0 from CRAN "
[42] "Installing \033[32mMatchLinReg\033[39m 0.7.0 from CRAN "
[43] "Successfully loaded lockbox"
```

davidboren commented:
I'll have to set the lockbox option to use a disposable, local lockbox directory, and it'll take a looooong-ass time to run when we do run the whole thing, but we'll get it.  A final, seeded test will also take a while to run, so we'll have to figure the specifics of that out later.

Commit:
Improve support of weird CRAN versions and make default download dir
safer

Commit:
Correct naming

davidboren commented:
Running into some interesting issues with CRAN packages with Bioconductor-only dependencies in their description files (In Imports rather than Depends).  What do we think our default behavior should be here?  When installing these packages with install.packages CRAN does not search Bioconductor.  Should we ignore them and install?  Prevent installation?  Should we add in Bioconductor logic (madness)?

Commit:
Ignore non-CRAN dependency packages

Commit:
Add compatibility for latest CRAN packages that can't be built without
author-specific hacks

davidboren commented:
Update on progress.  I've decided to ignore dependencies not on CRAN and issue a warning.  I've also moved all dependency installations to binary type, because some of these authors do wacky-ass things during compilation.  If a locked package fails during compilation I figure that is on the user, though I'm not opposed to the addition of a "type" entry in the lockfile logic, so that someone can specify binary if they are having trouble with their locked packages.  Thoughts?

Commit:
Make all dependencies binary installations

Commit:
Fix simplified download logic to handle NA version

Commit:
Source binaries if necessary

Commit:
Add missing argument

peterhurford commented:
@davidboren Does it work with remotes?

peterhurford commented:
@davidboren Let me know when you're done and I can review more.

Commit:
Always try to install source first

davidboren commented:
@peterhurford Yeah, remotes are supported, though if there are multiple appearances of the same package  (one package needs testthatsomemore, another needs the remote version of testthatsomemore) then it looks for the source that has the latest available version.  Version determines all.  That could be changed to github Remotes win every time also...

We can stop on the non-CRAN available dependencies, but we'll have to figure out what to do when running that sweeping test through random CRAN packages, since I found this issue because some were referencing non-cran packages.  Oddly, CRAN will still install and doesn't care...  I agree that for safety it'd be better to give a hard stop.  On the other hand that hard stop could just be given for Depends level dependencies rather than Imports level, I suppose.  That'd be work though.  I hate work.

Okay, so the logic on dependency installations is now such:
- If the package is a dependency AND the environment supports binaries (.Platform$pkgType), establish the latest available version as the latest BINARY version number.
- Otherwise, use the latest source version to establish the version.
- Download the source at this version, whichever it is

During installation:
- Try installing the source at this version.  If it fails, try to install the binary version.

This ensures that, even when we have issues with compilation because authors are crazy (like hardcoded test path creation crazy...I seen it), we'll still be able to install the binary at the same version level, making lockbox happy in that we still install the expected version.

Commit:
Make lockbox fully compatible with ridiculous dash-versioning schemes

davidboren commented:
@robertzk Another note, if we are serious about a connected Syberia landscape I think we should be setting Remotes in our description files right now, especially since we don't have a time-frame for releasing on CRAN.  I'm willing to set that up and put in pull requests on those items, since as of now in order for a lockfile to work with a subset of them we have to think about the dependencies very carefully because they aren't on cran and aren't specified in remotes, but nonetheless often depend on each-other through fields in the description files.

Commit:
Add test for full lockbox installation with random dependencies

Commit:
Update logic

Commit:
Rename test function.  Eliminate incomplete final line warning

Commit:
Suppress warnings around readlines.  Remove newline in success message

Commit:
Find out why travis is failing

Commit:
TryCatch random lockbox test

Commit:
Finalize test formatting

davidboren commented:
Miniature version of test (using just 1 random CRAN package right now along with 10 from syberia-verse...still not a super-fast test) currently passes using devtools::test() but fails using devtools::check().  I'm definitely missing something obvious about the environment R CMD check produces.  Tried to reproduce it by starting R with "R_DEFAULT_PACKAGES= LC_COLLATE=C  R" but test() is still successful.  If anyone has any ideas let me know.  I'll try to debug.

Commit:
Test workaround

davidboren commented:
Found a workaround here: https://github.com/hadley/testthat/issues/86
See if it pleases Travis.  Dinnertime.

davidboren commented:
Weirdly, the workaround fixes the check() run locally but not on Travis.  Good times.

Commit:
Default seed to NULL

Commit:
One last style change

Commit:
Error printing

Commit:
Correct error printing

Commit:
NULL default github_pat

Commit:
Add larger number of cran dependencies

Commit:
Check for package only available in source

davidboren commented:
@robertzk @peterhurford So I think doing X random CRAN packages is kind of a bust.  Just look at the the after failure portion of this Travis build to see what I mean: https://travis-ci.org/robertzk/lockbox/builds/114380263  RODBC has outside R dependencies, and fails when installing from source.  There are no binaries available for it (it requires compilation), and iODBC is required to be downloaded for installation: http://stackoverflow.com/questions/26210317/installation-of-rodbc-roracle-packages-on-os-x-mavericks  I suspect there are waaaaay more packages like this that require hands-on intervention, and I don't think it's feasible to expect dependency installation that can deal with outside R dependencies, especially when considering that there are Bioconductor dependencies mixed in with some of these CRAN packages and I agree with Peter that lockbox should stop at those packages rather than issue a warning.

Currently I've made it so that even when lockbox isn't in verbose mode it is still informative about these kinds of failures, and I honestly think that should be enough for most people.  The fact that it will attempt a binary install if the source fails should catch a lot of problems as well, but there is only so much that can be done.  I can hardcode in a yml of some CRAN packages for a test that passes Travis, but I think a random selection is out.  If this is too much stochastic weirdness for lockbox then I can just keep everything in my fork, which I am comfortable doing.

peterhurford commented:
@davidboren Still sounds like you learned a lot and improved the code.  How about you do a static test from a static lockfile?

davidboren commented:
Agreed.  Will just hardcode the cran yml

Commit:
Stop on unavailable dependencies and hardcode cran yml

Commit:
Set latest version from download only on dependency package

davidboren commented:
@peterhurford Okay, as far as I'm concerned this is as ready to go as it's going to be.  Using 10 syberia and 5 cran packages in the test (which is still not by any means a short test), and we're passing with no trouble.  Look over anything you want and let me know.

peterhurford commented:
@davidboren Looks good. The test is great.

The code could be cleaner but that sounds hard and we can refactor in a later PR...

The only other thing worth revisiting is maybe squishing the commit history up. That makes me nervous but if you feel confident doing it, that would be nice.

Otherwise I'm good to merge.

robertzk commented:
Read through the full conversation. I'm happy with the compromises. Agree that packages that require non-R dependencies or do funky things in the build process should be exempt from the end-to-end test. I'll merge this now.

robertzk commented:
Actually let me look over the [new changes](https://github.com/davidboren/lockbox/compare/cb384a61f00ff98072ed7dfc6d2ed664261f2c5b...davidboren:364f274ed73d0aca3821d6927b108d5975576ce5) first.

robertzk commented on R/library.R:
```diff
@@ -191,11 +167,116 @@ current_version.locked_package <- function(package) {
+  if (is.null(ref)) arguments <- list(package$repo)
+  else arguments <- list(paste(package$repo, ref, sep = "@"))
+
+  token <- Sys.getenv("GITHUB_PAT")
```
Can you use `github_pat()` for this please?

robertzk commented on R/lockbox.R:
```diff
@@ -14,48 +14,55 @@
 
 #' @export
-lockbox.character <- function(file, env) {
+lockbox.character <- function(file, env = getOption("lockbox.env", "!packages")) {
```
You should never need to replicate default parameters in an S3 specialization method.

davidboren replied:
Weirdly, I was noticing that the env wasn't getting passed to the type-specific functions... was banging my head against the wall.  Will check in your master to see if I can prove it was so :)

davidboren replied:
Definitely not so in master.  will see if I can eliminate this issue in my fork

davidboren replied:
Fixed.  No idea why it cropped up in the first place.

robertzk commented on R/lockbox.R:
```diff
@@ -68,22 +75,38 @@ lockbox.default <- function(obj) {
+}
+
+set_download_dir <- function() {
+  download_dir <- lockbox_download_dir()
```
Weird--there already is a function `lockbox_download_dir`. So why do you have a `lockbox_library` function above that seems to do the same thing?

robertzk commented:
Few minor comments, otherwise looks excellent. :100:

Commit:
use github_pat function.  Fix env default assignment.Change download
directory

davidboren commented:
Definitely wait until it passes the test again, but I've taken care of those last comments.  I've got to run.  Picking up a drugged up dog from surgery.  Will check back in later this evening in case something cropped up.

robertzk commented:
Oh my bad... I think `github_pat` is in devtools not lockbox! Carry on -- or make the helper function.

Commit:
Handle PAT correctly

davidboren commented:
I just made the helper function.  Test is green again after I broke things with said helper function.  So the test is useful.

I think I've addressed the comments.  Should I squash this thing?  Let me know.
